File: app/src/main/java/com/example/taxconnect/MyDocumentsActivity.java
package com.example.taxconnect;

import android.os.Bundle;
import android.view.View;
import android.widget.Toast;
import android.app.DownloadManager;
import android.content.Context;
import android.net.Uri;
import android.os.Environment;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.taxconnect.databinding.ActivityMyDocumentsBinding;
import com.example.taxconnect.services.CloudinaryHelper;
import com.google.firebase.auth.FirebaseAuth;

import com.example.taxconnect.adapter.DocumentAdapter;
import com.example.taxconnect.adapter.FolderAdapter;
import com.example.taxconnect.model.DocumentModel;
import com.example.taxconnect.model.UserModel;
import com.google.firebase.Timestamp;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import android.widget.ArrayAdapter;

public class MyDocumentsActivity extends AppCompatActivity {

    private ActivityMyDocumentsBinding binding;
    private DocumentAdapter adapter;
    private FolderAdapter folderAdapter;
    private List<DocumentModel> allDocuments = new ArrayList<>();
    private String currentFolder = "All";

    private final ActivityResultLauncher<String> pickDocLauncher = registerForActivityResult(
            new ActivityResultContracts.GetContent(),
            uri -> {
                if (uri != null) {
                    uploadDocument(uri);
                }
            }
    );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMyDocumentsBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setupToolbar();
        setupFolderRecyclerView();
        setupRecyclerView();
        setupListeners();
        checkUserRoleAndCustomizeUI();
        loadDocuments();
    }

    private void setupFolderRecyclerView() {
        List<String> folders = new ArrayList<>();
        folders.add("All");
        folders.add("Personal");
        folders.add("Business");
        folders.add("FY 2023-24");
        folders.add("FY 2024-25");
        folders.add("Others");

        folderAdapter = new FolderAdapter(folders, currentFolder, folder -> {
            currentFolder = folder;
            filterDocuments();
        });
        
        binding.rvFolders.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));
        binding.rvFolders.setAdapter(folderAdapter);
    }

    private void filterDocuments() {
        if (currentFolder.equals("All")) {
            adapter.setDocuments(allDocuments);
            binding.tvEmptyState.setVisibility(allDocuments.isEmpty() ? View.VISIBLE : View.GONE);
        } else {
            List<DocumentModel> filtered = new ArrayList<>();
            for (DocumentModel doc : allDocuments) {
                if (currentFolder.equals(doc.getCategory()) || 
                   (doc.getCategory() == null && "Uncategorized".equals(currentFolder))) {
                    filtered.add(doc);
                }
            }
            adapter.setDocuments(filtered);
            binding.tvEmptyState.setVisibility(filtered.isEmpty() ? View.VISIBLE : View.GONE);
        }
    }

    private void checkUserRoleAndCustomizeUI() {
        String uid = FirebaseAuth.getInstance().getUid();
        if (uid == null) return;

        DataRepository.getInstance().fetchUser(uid, new DataRepository.DataCallback<UserModel>() {
            @Override
            public void onSuccess(UserModel user) {
                if (user != null && !"CA".equals(user.getRole())) {
                    // Customer View
                    binding.toolbarDocs.setTitle("My Documents");
                    binding.tvUploadTitle.setText("Upload Documents");
                    binding.tvUploadDesc.setText("Keep your tax documents and identity proofs organized and secure.");
                    binding.tvListTitle.setText("Uploaded Documents");
                    binding.btnPickDoc.setText("Upload Document");
                }
            }

            @Override
            public void onError(String error) {
                // Default to CA view or existing view on error
            }
        });
    }

    private void setupToolbar() {
        binding.toolbarDocs.setNavigationOnClickListener(v -> finish());
    }

    private void setupListeners() {
        binding.btnPickDoc.setOnClickListener(v -> pickDocLauncher.launch("*/*"));
    }

    private void setupRecyclerView() {
        adapter = new DocumentAdapter(document -> {
            try {
                DownloadManager.Request request = new DownloadManager.Request(Uri.parse(document.getUrl()));
                request.setTitle(document.getName());
                request.setDescription("Downloading file...");
                request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
                request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, document.getName());

                DownloadManager manager = (DownloadManager) getSystemService(Context.DOWNLOAD_SERVICE);
                if (manager != null) {
                    manager.enqueue(request);
                    Toast.makeText(this, "Download started", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(this, "Download failed: Manager not available", Toast.LENGTH_SHORT).show();
                }
            } catch (Exception e) {
                Toast.makeText(this, "Download error: " + e.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
        binding.rvMyDocs.setLayoutManager(new LinearLayoutManager(this));
        binding.rvMyDocs.setAdapter(adapter);
    }

    private void loadDocuments() {
        String uid = FirebaseAuth.getInstance().getUid();
        if (uid == null) return;

        DataRepository.getInstance().getDocuments(uid, new DataRepository.DataCallback<List<DocumentModel>>() {
            @Override
            public void onSuccess(List<DocumentModel> data) {
                allDocuments = data;
                // Retrofit old documents
                for (DocumentModel doc : allDocuments) {
                    if (doc.getCategory() == null) {
                        doc.setCategory("Uncategorized");
                    }
                }
                filterDocuments();
            }

            @Override
            public void onError(String error) {
                Toast.makeText(MyDocumentsActivity.this, "Failed to load docs", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void uploadDocument(android.net.Uri uri) {
        binding.uploadProgress.setVisibility(View.VISIBLE);
        binding.btnPickDoc.setEnabled(false);

        // Get file name
        String fileName = getFileName(uri);

        CloudinaryHelper.uploadMedia(this, uri, new CloudinaryHelper.ImageUploadCallback() {
            @Override
            public void onSuccess(String url) {
                runOnUiThread(() -> {
                    // Save to Firestore
                    saveDocumentToFirestore(fileName, url);
                });
            }

            @Override
            public void onError(String error) {
                runOnUiThread(() -> {
                    binding.uploadProgress.setVisibility(View.GONE);
                    binding.btnPickDoc.setEnabled(true);
                    Toast.makeText(MyDocumentsActivity.this, "Upload failed: " + error, Toast.LENGTH_SHORT).show();
                });
            }
        });
    }

    private void saveDocumentToFirestore(String name, String url) {
        String uid = FirebaseAuth.getInstance().getUid();
        if (uid == null) return;

        DocumentModel doc = new DocumentModel();
        doc.setId(java.util.UUID.randomUUID().toString());
        doc.setName(name);
        doc.setUrl(url);
        doc.setType(name.endsWith(".pdf") ? "pdf" : "image");
        String category = "Uncategorized";
        if (!currentFolder.equals("All")) {
            category = currentFolder;
        }
        doc.setCategory(category);
        doc.setUploadedAt(Timestamp.now());

        DataRepository.getInstance().saveDocument(uid, doc, new DataRepository.DataCallback<Boolean>() {
            @Override
            public void onSuccess(Boolean data) {
                binding.uploadProgress.setVisibility(View.GONE);
                binding.btnPickDoc.setEnabled(true);
                Toast.makeText(MyDocumentsActivity.this, "Document Uploaded", Toast.LENGTH_SHORT).show();
                loadDocuments(); // Reload list
            }

            @Override
            public void onError(String error) {
                binding.uploadProgress.setVisibility(View.GONE);
                binding.btnPickDoc.setEnabled(true);
                Toast.makeText(MyDocumentsActivity.this, "Save failed: " + error, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private String getFileName(android.net.Uri uri) {
        String result = null;
        if (uri.getScheme().equals("content")) {
            try (android.database.Cursor cursor = getContentResolver().query(uri, null, null, null, null)) {
                if (cursor != null && cursor.moveToFirst()) {
                    int index = cursor.getColumnIndex(android.provider.OpenableColumns.DISPLAY_NAME);
                    if (index >= 0) {
                        result = cursor.getString(index);
                    }
                }
            }
        }
        if (result == null) {
            result = uri.getPath();
            int cut = result.lastIndexOf('/');
            if (cut != -1) {
                result = result.substring(cut + 1);
            }
        }
        return result;
    }
}

File: app/src/main/java/com/example/taxconnect/features/documents/MyDocumentsActivity.kt
package com.example.taxconnect.features.documents

import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.view.View
import android.view.LayoutInflater
import androidx.activity.result.contract.ActivityResultContracts
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.taxconnect.R
import com.example.taxconnect.features.documents.DocumentAdapter
import com.example.taxconnect.features.documents.FolderAdapter
import com.example.taxconnect.core.base.BaseActivity
import com.example.taxconnect.data.repositories.DataRepository
import com.example.taxconnect.data.repositories.DataRepository.DataCallback
import com.example.taxconnect.databinding.ActivityMyDocumentsBinding
import com.example.taxconnect.data.models.DocumentModel
import com.example.taxconnect.data.models.UserModel
import com.example.taxconnect.data.services.CloudinaryHelper
import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import java.util.UUID
import com.example.taxconnect.core.cache.DocumentCacheManager
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import androidx.lifecycle.lifecycleScope

class MyDocumentsActivity : BaseActivity<ActivityMyDocumentsBinding>() {

    override val bindingInflater: (LayoutInflater) -> ActivityMyDocumentsBinding = ActivityMyDocumentsBinding::inflate
    private lateinit var adapter: DocumentAdapter
    private lateinit var folderAdapter: FolderAdapter
    private var allDocuments: List<DocumentModel> = ArrayList()
    private var currentFolder = "All"

    private val pickDocLauncher = registerForActivityResult(ActivityResultContracts.GetContent()) { uri ->
        if (uri != null) {
            uploadDocument(uri)
        }
    }

    override fun initViews() {
        setupToolbar()
        setupFolderRecyclerView()
        setupRecyclerView()
        checkUserRoleAndCustomizeUI()
        loadDocuments()
    }

    override fun observeViewModel() {
        // No ViewModel to observe
    }

    private var selectedUploadCategory: String? = null

    override fun setupListeners() {
        binding.btnPickDoc.setOnClickListener {
            val bottomSheet = UploadDocumentBottomSheet { category ->
                selectedUploadCategory = category
                pickDocLauncher.launch("*/*")
            }
            bottomSheet.show(supportFragmentManager, UploadDocumentBottomSheet.TAG)
        }
    }

    private fun setupFolderRecyclerView() {
        val folders = listOf("All", "Personal", "Business", "FY 2023-24", "FY 2024-25", "Others")

        folderAdapter = FolderAdapter(folders, currentFolder, object : FolderAdapter.OnFolderClickListener {
            override fun onFolderClick(folder: String) {
                currentFolder = folder
                filterDocuments()
            }
        })

        binding.rvFolders.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        binding.rvFolders.adapter = folderAdapter
    }

    private fun filterDocuments() {
        val filtered = if (currentFolder == "All") {
            allDocuments
        } else {
            allDocuments.filter { doc ->
                currentFolder == doc.category || (doc.category == null && "Uncategorized" == currentFolder)
            }
        }
        adapter.setDocuments(filtered)
        
        if (filtered.isEmpty()) {
            binding.layoutEmptyState.root.visibility = View.VISIBLE
            binding.layoutEmptyState.tvEmptyTitle.text = "No Documents"
            binding.layoutEmptyState.tvEmptyDescription.text = "You haven't uploaded any documents in this folder yet."
            binding.layoutEmptyState.ivEmptyIcon.setImageResource(R.drawable.ic_document)
            binding.layoutEmptyState.btnEmptyAction.text = "Upload Document"
            binding.layoutEmptyState.btnEmptyAction.setOnClickListener { binding.btnPickDoc.performClick() }
        } else {
            binding.layoutEmptyState.root.visibility = View.GONE
        }
    }

    private fun checkUserRoleAndCustomizeUI() {
        val uid = FirebaseAuth.getInstance().uid ?: return

        DataRepository.getInstance().fetchUser(uid, object : DataCallback<UserModel> {
            override fun onSuccess(data: UserModel?) {
                if (data != null && "CA" != data.role) {
                    // Customer View
                    binding.toolbarDocs.title = getString(R.string.my_documents)
                    binding.tvUploadTitle.text = getString(R.string.upload_documents)
                    binding.tvUploadDesc.text = getString(R.string.upload_documents_desc)
                    binding.tvListTitle.text = getString(R.string.uploaded_documents)
                    binding.btnPickDoc.text = getString(R.string.upload_documents)
                }
            }

            override fun onError(error: String?) {
                // Default to CA view or existing view on error
            }
        })
    }

    private fun setupToolbar() {
        binding.toolbarDocs.setNavigationOnClickListener { finish() }
    }

    private fun setupRecyclerView() {
        adapter = DocumentAdapter(object : DocumentAdapter.OnDocumentClickListener {
            override fun onDownloadClick(document: DocumentModel) {
                handleDocumentDownload(document)
            }

            override fun onViewClick(document: DocumentModel) {
                openDocument(document)
            }

            override fun onCacheStatusChanged() {
                // Refresh the list to update cache indicators
                adapter.notifyDataSetChanged()
            }
        })
        binding.rvMyDocs.layoutManager = LinearLayoutManager(this)
        binding.rvMyDocs.adapter = adapter
    }

    private fun loadDocuments() {
        val uid = FirebaseAuth.getInstance().uid ?: return

        binding.shimmerViewContainer.root.visibility = View.VISIBLE
        binding.shimmerViewContainer.root.startShimmer()
        binding.layoutEmptyState.root.visibility = View.GONE
        binding.rvMyDocs.visibility = View.GONE
        binding.layoutErrorState.root.visibility = View.GONE

        DataRepository.getInstance().getDocuments(uid, object : DataCallback<List<DocumentModel>> {
            override fun onSuccess(data: List<DocumentModel>?) {
                binding.shimmerViewContainer.root.stopShimmer()
                binding.shimmerViewContainer.root.visibility = View.GONE
                
                if (data != null) {
                    allDocuments = data
                    // Retrofit old documents
                    for (doc in allDocuments) {
                        if (doc.category == null) {
                            doc.category = getString(R.string.uncategorized)
                        }
                    }
                    filterDocuments()
                }
            }

            override fun onError(error: String?) {
                binding.shimmerViewContainer.root.stopShimmer()
                binding.shimmerViewContainer.root.visibility = View.GONE
                
                if (allDocuments.isEmpty()) {
                     binding.rvFolders.visibility = View.GONE
                     binding.rvMyDocs.visibility = View.GONE
                     binding.layoutEmptyState.root.visibility = View.GONE
                     
                     val errorBinding = binding.layoutErrorState
                     errorBinding.root.visibility = View.VISIBLE
                     errorBinding.tvErrorTitle.text = "Error Loading Documents"
                     errorBinding.tvErrorDescription.text = error ?: "Unknown error"
                     errorBinding.btnRetry.setOnClickListener {
                         errorBinding.root.visibility = View.GONE
                         loadDocuments()
                     }
                } else {
                    showToast(error ?: getString(R.string.failed_to_load_docs))
                }
            }
        })
    }

    private fun uploadDocument(uri: Uri) {
        binding.uploadProgress.visibility = View.VISIBLE
        binding.btnPickDoc.isEnabled = false

        val fileName = getFileName(uri)

        CloudinaryHelper.uploadMedia(this, uri, object : CloudinaryHelper.ImageUploadCallback {
            override fun onSuccess(data: String?) {
                runOnUiThread {
                    saveDocumentToFirestore(fileName, data ?: "")
                }
            }

            override fun onError(error: String?) {
                runOnUiThread {
                    binding.uploadProgress.visibility = View.GONE
                    binding.btnPickDoc.isEnabled = true
                    showToast(getString(R.string.upload_failed, error))
                }
            }
        })
    }

    private fun saveDocumentToFirestore(name: String?, url: String) {
        val uid = FirebaseAuth.getInstance().uid ?: return

        val doc = DocumentModel()
        doc.id = UUID.randomUUID().toString()
        doc.name = name
        doc.url = url
        doc.type = if (name?.endsWith(".pdf", ignoreCase = true) == true) "pdf" else "image"
        doc.category = selectedUploadCategory ?: (if (currentFolder == "All") "Uncategorized" else currentFolder)
        doc.uploadedAt = Timestamp.now()

        DataRepository.getInstance().saveDocument(uid, doc, object : DataCallback<Boolean> {
            override fun onSuccess(data: Boolean?) {
                binding.uploadProgress.visibility = View.GONE
                binding.btnPickDoc.isEnabled = true
                showToast(getString(R.string.document_uploaded))
                loadDocuments()
            }

            override fun onError(error: String?) {
                binding.uploadProgress.visibility = View.GONE
                binding.btnPickDoc.isEnabled = true
                showToast(getString(R.string.save_failed, error))
            }
        })
    }

    private fun getFileName(uri: Uri): String? {
        var result: String? = null
        if (uri.scheme == "content") {
            contentResolver.query(uri, null, null, null, null)?.use { cursor ->
                if (cursor.moveToFirst()) {
                    val index = cursor.getColumnIndex(android.provider.OpenableColumns.DISPLAY_NAME)
                    if (index >= 0) {
                        result = cursor.getString(index)
                    }
                }
            }
        }
        if (result == null) {
            result = uri.path
            val cut = result?.lastIndexOf('/') ?: -1
            if (cut != -1) {
                result = result?.substring(cut + 1)
            }
        }
        return result
    }


    private fun handleDocumentDownload(document: DocumentModel) {
        val url = document.url ?: return
        
        // Check if already cached
        if (DocumentCacheManager.isDocumentCached(this, url)) {
            Snackbar.make(binding.root, getString(R.string.available_offline), Snackbar.LENGTH_SHORT).show()
            return
        }
        
        // Download and cache
        binding.uploadProgress.visibility = View.VISIBLE
        Snackbar.make(binding.root, getString(R.string.downloading), Snackbar.LENGTH_SHORT).show()
        
        lifecycleScope.launch {
            val result = DocumentCacheManager.cacheDocument(this@MyDocumentsActivity, url)
            
            withContext(Dispatchers.Main) {
                binding.uploadProgress.visibility = View.GONE
                
                if (result.isSuccess) {
                    Snackbar.make(binding.root, getString(R.string.download_complete), Snackbar.LENGTH_LONG).show()
                    adapter.notifyDataSetChanged() // Refresh to show offline indicator
                } else {
                    Snackbar.make(binding.root, getString(R.string.download_failed), Snackbar.LENGTH_LONG).show()
                }
            }
        }
    }

    private fun openDocument(document: DocumentModel) {
        val url = document.url ?: return
        
        // Check if cached, use cached version
        val cachedFile = DocumentCacheManager.getCachedDocument(this, url)
        val docUrl = if (cachedFile != null) {
            Uri.fromFile(cachedFile).toString()
        } else {
            url
        }
        
        val intent = Intent(this@MyDocumentsActivity, SecureDocViewerActivity::class.java)
        intent.putExtra("DOC_URL", docUrl)
        intent.putExtra("DOC_TITLE", document.name)
        var type = "image"
        if (document.name?.lowercase()?.endsWith(".pdf") == true) {
            type = "pdf"
        }
        intent.putExtra("DOC_TYPE", type)
        startActivity(intent)
    }
}
